chalice/deploy/newdeployer.py

```diff
@@ -198,6 +199,27 @@ def build(self, config, stage_name):
 
     def _get_role_reference(self, config, stage_name, function):
         # type: (Config, str, app.LambdaFunction) -> models.IAMRole
+        role = self._create_role_reference(config, stage_name, function)
+        role_identifier = self._get_role_identifier(role)
+        if role_identifier in self._known_roles:
+            # If we've already create a models.IAMRole with the same
+            # identifier, we'll use the existing object instead of
+            # creating a new one.
+            return self._known_roles[role_identifier]
+        self._known_roles[role_identifier] = role
+        return role
+
+    def _get_role_identifier(self, role):
+        # type: (models.IAMRole) -> str
+        if isinstance(role, models.PreCreatedIAMRole):
+            return role.role_arn
+        # We know that if it's not a PreCreatedIAMRole, it's
+        # a managed role, so we're using cast() to make mypy happy.
+        role = cast(models.ManagedIAMRole, role)
```

Comment:
An `IAMRole` doesn't have a `resource_name`, which is likely what it was complaining about. The cast kind of irks me, perhaps an `elif isinstance(role, models.ManagedIAMRole)` would be better?

Reply:
Yep, I can do the `elif` which would satisfy mypy.  The tradeoff here is that coverage will complain that you're missing branch coverage where both if statements are false, that is, it's neither type `PreCreatedIAMRole` nor `ManagedIAMRole`.  In this case we know it's impossible for that to happen because those are the only subclasses of `IAMRole`, so if you guys are fine with codecov complaining I can switch this to `elif`.

Reply:
I don't have a strong preference especially since refactoring will probably remove this line prior to merging.

Reply:
Another possible solution is to change the type signature from taking an `IAMRole` to `Union[PreCreatedIAMRole, ManagedIAMRole]`. If we're certain this will only take these two types we should let mypy know that. We can then implement the second case as just an `else` which should avoid coverage issues as well.

Reply:
This does have another trade off for functions that produce `IAMRole`. It is unfortunate mypy doesn't know all the concrete instances.

Reply:
Yeah I would have to update every function/method that returned type `IAMRole` which I don't think is worth it.

It also turns out the `elif` isn't enough still.  Mypy also detects that it's possible to return a non str type if neither of the branches are true:

```python
    def _get_role_identifier(self, role):
        # type: (models.IAMRole) -> str
        if isinstance(role, models.PreCreatedIAMRole):
            return role.role_arn
        elif isinstance(role, models.ManagedIAMRole):
            return role.resource_name
```

```
$ make typecheck
mypy --py2 --ignore-missing-imports --follow-imports=skip -p chalice --disallow-untyped-defs --strict-optional --warn-no-return
chalice/deploy/newdeployer.py:212: error: Missing return statement
make: *** [typecheck] Error 1
```

It seems like the `cast()` is the most reasonable option so far.  Thoughts?